chromium_src/chrome/browser/ssl/https_only_mode_navigation_throttle.cc

```diff
@@ -0,0 +1,143 @@
+#include "chrome/browser/ssl/https_only_mode_controller_client.h"
+#include "components/prefs/pref_service.h"
+#include "components/security_interstitials/content/stateful_ssl_host_state_delegate.h"
+#include "content/public/browser/navigation_handle.h"
+#include "content/public/browser/page_navigator.h"
+#include "content/public/browser/web_contents.h"
+
+namespace content {
+class BrowserContext;
+}  // namespace content
+
+namespace {
+
+// Tor is slow and needs a longer fallback delay
+constexpr base::TimeDelta kTorFallbackDelay = base::Seconds(20);
+
+bool ShouldUpgradeToHttps(content::NavigationHandle* handle) {
+  if (!brave_shields::IsHttpsByDefaultFeatureEnabled()) {
+    return false;
+  }
+  content::BrowserContext* context =
+      handle->GetWebContents()->GetBrowserContext();
+  const GURL& url = handle->GetURL();
+  HostContentSettingsMap* map =
+      HostContentSettingsMapFactory::GetForProfile(context);
+  return brave_shields::ShouldUpgradeToHttps(
+      map, url, g_brave_browser_process->https_upgrade_exceptions_service());
+}
+
+bool IsTor(content::NavigationHandle* handle) {
+  auto* context = handle->GetWebContents()->GetBrowserContext();
+  Profile* profile = Profile::FromBrowserContext(context);
+  return profile->IsTor();
+}
+
+}  // namespace
+
+#define WillFailRequest WillFailRequest_ChromiumImpl
+#define GetBoolean(PREF_NAME) \
+  GetBooleanOr(PREF_NAME, ShouldUpgradeToHttps(handle))
+#define SetNavigationTimeout(DEFAULT_TIMEOUT)                         \
+  SetNavigationTimeout(IsTor(navigation_handle()) ? kTorFallbackDelay \
+                                                  : DEFAULT_TIMEOUT)
+
+#include "src/chrome/browser/ssl/https_only_mode_navigation_throttle.cc"
+
+#undef WillFailRequest
+#undef GetBoolean
+#undef SetNavigationTimeout
+
+// Called if there is a non-OK net::Error in the completion status.
+content::NavigationThrottle::ThrottleCheckResult
+HttpsOnlyModeNavigationThrottle::WillFailRequest() {
+  // Don't fall back if the HTTPS by Default feature flag
+  // is disabled.
+  if (!brave_shields::IsHttpsByDefaultFeatureEnabled()) {
+    return WillFailRequest_ChromiumImpl();
+  }
+
+  auto* handle = navigation_handle();
+  auto* contents = handle->GetWebContents();
+  const GURL& request_url = handle->GetURL();
+
+  // Don't fall back if we are using HTTPS-Only Mode or the HTTPS-Upgrade
+  // Strict setting is enabled.
+  content::BrowserContext* context =
+      handle->GetWebContents()->GetBrowserContext();
+  HostContentSettingsMap* map =
+      HostContentSettingsMapFactory::GetForProfile(context);
+  Profile* profile = Profile::FromBrowserContext(context);
+  auto* prefs = profile->GetPrefs();
+  if ((prefs && prefs->GetBoolean(prefs::kHttpsOnlyModeEnabled)) ||
+      brave_shields::ShouldForceHttps(map, request_url)) {
+    return WillFailRequest_ChromiumImpl();
+  }
+
+  // Fall back only on TLS-related errors.
+  const net::SSLInfo info = handle->GetSSLInfo().value_or(net::SSLInfo());
```

Comment:
`value_or(net::SSLInfo())` is concerning.

please consider this:
```
if (handle->GetSSLInfo() && !net::IsCertStatusError(handle->GetSSLInfo()->cert_status) && ...
```

Reply:
fixed